### ISSUE TITLE
Feature Request: Expose monitoring metrics through prometheus endpoint

### ISSUE THREAD

**Issue by SomniVertix** [TechnitiumSoftware/DnsServer#525]
Description: Would like to be able to scrape the metrics that are shown in the dashboard as prometheus metrics

**Comment by ShreyasZare** [comment]
Thanks for the feature request. I don't any any experience working with it so not sure if I would be able to accommodate it at the moment. I may try it later once all of the planned features are in place.

**Comment by CRASH-Tech** [comment]
Up. this is very necessary feature for production use.

**Comment by danthonywalker** [comment]
I also vote for this feature.

Since this project is in C#, you'd use https://github.com/prometheus-net/prometheus-net and default metrics would at least expose important things such as runtime processes.

https://github.com/NathanPERIER/technitium-dns-prometheus-exporter gives a good idea of what custom metrics would be useful to expose.

**Comment by vakev** [comment]
+1 
It will be quite useful.

**Comment by IngmarStein** [comment]
@ShreyasZare would you accept a PR to add a metrics endpoint for Prometheus?

**Comment by ShreyasZare** [comment]
> [@ShreyasZare](https://github.com/ShreyasZare) would you accept a PR to add a metrics endpoint for Prometheus?

@IngmarStein Thanks for asking. There was a discussion on issue #1664 recently and I have some basic API for metric implemented in local dev code that returns a bunch of counters. Not sure if that is enough but, it can be expanded with more data. If I can get exact requirements for this API then it would be much easy for me to add support for it. Right now, I do not have any experience with Prometheus and do not really have time to spend to find out how it works and what data will be required by it.

**Comment by IngmarStein** [comment]
Yeah, your post in #1664 reminded me that it would be much nicer to have Prometheus support built-in so that it doesn't require a sidecar process just for that.

I think you've already done most of the work, the data just needs to be made available in a format understandable by Prometheus. The most common one is also rather easy to produce: it's a line-oriented text format with one line per metric (which can include tags) and annotations for a human-readable help text (which describes the metric) and the type.
The spec is at https://docs.google.com/document/u/1/d/1ZjyKiKxZV83VI9ZKAXRGKaUKK2BIWCT7oiGBKDBpjEY, but here's a tl;dr: https://o11y.eu/blog/prometheus-text-format/.

It is conventional to make this data available at `/metrics`

@danthonywalker already linked to https://github.com/prometheus-net/prometheus-net which would add some standard metrics common to .NET applications, but using it is not strictly a must. You'd not have to worry about the data format or setting the right content-type when using such a library.

**Comment by ShreyasZare** [comment]
@IngmarStein Thanks for the links. Is the text format preferred or does JSON works? I guess JSON is older and thus has better support. It would be easier to implement it since the API already handles JSON.

**Comment by IngmarStein** [comment]
@ShreyasZare Prometheus can't process JSON natively. Prometheus is designed to scrape metrics in its own plain-text exposition format (or OpenMetrics). It cannot parse a random JSON response on its own. A workaround is another proxy process that translates JSON to the Prometheus format but that is functionally equivalent to https://github.com/NathanPERIER/technitium-dns-prometheus-exporter.

**Comment by ShreyasZare** [comment]
@IngmarStein The documentation link you shared also defines JSON format. Is it not supported?

**Comment by IngmarStein** [comment]
The doc says "The text format replaces the JSON format (version 0.0.2 and 0.0.1) as the human readable format. The JSON format is now deprecated, does not support all features of client libraries anymore, and will be removed eventually." The doc is from 2014, so I assume it's no longer there (but I haven't checked).

**Comment by ShreyasZare** [comment]
Thanks for the clarification. Will see how this can be implemented.

**Comment by ShreyasZare** [comment]
Technitium DNS Server v15.0 is now available which includes support Prometheus with new API end point. The current implementation provides basic data and can be updated in later releases to add more stats. Do upgrade and let me know your feedback.

**Comment by IngmarStein** [comment]
For those following along: Technitium v15 doesn't use the standard /metrics location for the metrics endpoint and requires authentication, but it should work with a configuration like this:

```
metrics_path: "/api/dashboard/metrics/text"
authorization:
  credentials: "${token}"
```

Where a non-expiring token can be created with `curl http://${technitium}/api/user/createToken?user=XXX&pass=XXX&tokenName=Prometheus`

**Comment by IngmarStein** [comment]
I think we'll have to wait until 15.0.1 for Prometheus to successfully scrape the endpoint: https://github.com/TechnitiumSoftware/DnsServer/pull/1846
